pkg/webhook/main.go

```diff
@@ -0,0 +1,183 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package webhook
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/spf13/cobra"
+
+	v1 "k8s.io/api/admission/v1"
+	"k8s.io/api/admission/v1beta1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog"
+	// TODO: try this library to see if it generates correct json patch
+	// https://github.com/mattbaird/jsonpatch
+)
+
+var (
+	certFile string
+	keyFile  string
+	port     int
+)
+
+// CmdWebhook is used by agnhost Cobra.
+var CmdWebhook = &cobra.Command{
+	Use:   "webhook",
+	Short: "Starts a HTTP server, useful for testing MutatingAdmissionWebhook and ValidatingAdmissionWebhook",
+	Long: `Starts a HTTP server, useful for testing MutatingAdmissionWebhook and ValidatingAdmissionWebhook.
+After deploying it to Kubernetes cluster, the Administrator needs to create a ValidatingWebhookConfiguration
+in the Kubernetes cluster to register remote webhook admission controllers.`,
+	Args: cobra.MaximumNArgs(0),
+	Run:  main,
+}
+
+func init() {
+	CmdWebhook.Flags().StringVar(&certFile, "tls-cert-file", "",
+		"File containing the default x509 Certificate for HTTPS. (CA cert, if any, concatenated after server cert).")
+	CmdWebhook.Flags().StringVar(&keyFile, "tls-private-key-file", "",
+		"File containing the default x509 private key matching --tls-cert-file.")
+	CmdWebhook.Flags().IntVar(&port, "port", 443,
+		"Secure port that the webhook listens on")
+	CmdWebhook.MarkFlagRequired("tls-cert-file")
+	CmdWebhook.MarkFlagRequired("tls-private-key-file")
+}
+
+// admitv1beta1Func handles a v1beta1 admission
+type admitv1beta1Func func(v1beta1.AdmissionReview) *v1beta1.AdmissionResponse
+
+// admitv1beta1Func handles a v1 admission
+type admitv1Func func(v1.AdmissionReview) *v1.AdmissionResponse
+
+// admitHandler is a handler, for both validators and mutators, that supports multiple admission review versions
+type admitHandler struct {
+	v1beta1 admitv1beta1Func
+	v1      admitv1Func
+}
+
+func newDelegateToV1AdmitHandler(f admitv1Func) admitHandler {
+	return admitHandler{
+		v1beta1: delegateV1beta1AdmitToV1(f),
+		v1:      f,
+	}
+}
+
+func delegateV1beta1AdmitToV1(f admitv1Func) admitv1beta1Func {
+	return func(review v1beta1.AdmissionReview) *v1beta1.AdmissionResponse {
+		in := v1.AdmissionReview{Request: convertAdmissionRequestToV1(review.Request)}
+		out := f(in)
+		return convertAdmissionResponseToV1beta1(out)
+	}
+}
+
+// serve handles the http portion of a request prior to handing to an admit
+// function
+func serve(w http.ResponseWriter, r *http.Request, admit admitHandler) {
+	var body []byte
+	if r.Body != nil {
+		if data, err := ioutil.ReadAll(r.Body); err == nil {
+			body = data
+		}
```

Comment:
err != nil?

Reply:
It's copying the data from r.Body into body if no error occurred while reading. I don't see how it would make sense to check if err is not nil.